text/2017-08-30-simplify-defer.md

```diff
@@ -0,0 +1,182 @@
+        return Some(ptr::read(&h.value));
+    },
+    Err(h) => head = h,
+}
+```
+
+Calling the destructor will now come with some performance cost due to additional boxing
+(`defer` stores `FnOnce()` in a `Box`).
+Moreover, deferring destruction this way is a bit more unergonomic.
+Let's fix both problems...
+
+# Detailed design
+
+### `Deferred`
+
+Instead of having the triplet `defer_free`/`defer_drop`/`defer`, we'll have only `defer`
+with the same interface, but this time as a safe function:
+
+```rust
+impl Scope {
+    // Deferred execution of arbitrary function `f`.
+    fn defer<F: FnOnce() + Send + 'static>(&self, f: F);
+
+    // ...
+}
+```
+
+In most cases `defer` will be used to just deallocate memory or execute simple destruction
+routines. The passed `FnOnce()` closure will almost always fit into, say, 3 words.
+
+I have an [implementation](https://gist.github.com/stjepang/00d63b8febc07b297eae3480e80d0e91)
+of `Deferred` ready, which is able to store a small closure within itself, falling back
+to heap allocation for big closures.
+
+**TL;DR:** `Deferred` is to `FnOnce() + Send + 'static` what `SmallVec` is to `Vec`.
+
+### `Owned::from_ptr`
+
+Finally, let's reconsider the piece of code from the motivation section and make it more ergonomic.
+There are several unergonomic obstacles to destruction of `head`:
+
+1. A `Ptr<'scope>` cannot be passed to a `'static` closure.
+2. A raw pointer cannot be passed to a `Send` closure (a workaround is to convert the pointer to `usize`).
+3. Finally, the raw pointer must be passed to `Box::from_raw`.
+
+To simplify matters, let's introduce a new unsafe constructor, `Owned::from_ptr`:
+
+```rust
+match self.head.compare_and_set_weak(head, next, AcqRel, scope) {
+    Ok(()) => unsafe {
+        let head = Owned::from_ptr(head);
+        scope.defer(move || drop(head));
+        return Some(ptr::read(&h.value));
+    },
+    Err(h) => head = h,
+}
+```
+
+Strictly speaking, this code is possibly incorrect because an `Owned` and `Ptr` pointing to the
+same thing exist at the same time. This is okay because the `Owned` will not be used until the
+closure executes, but **@RalfJung**'s new unsafe code checker might reject the code as unsound.
+
+### `Ptr::to_static`
+
+We can fix the problem by introducing another unsafe helper method, `Ptr::to_static`, which
+converts `Ptr<'scope>` to `Ptr<'static>`:
+
+```rust
+match self.head.compare_and_set_weak(head, next, AcqRel, scope) {
+    Ok(()) => unsafe {
+        let head = head.to_static();
+        scope.defer(|| drop(Owned::from_ptr(head)));
+        return Some(ptr::read(&h.value));
+    },
+    Err(h) => head = h,
+}
+```
+
+# Drawbacks
+
+We'll introduce two new functions (`Owned::from_ptr` and `Ptr::to_static`). But at the same time
+we'll also remove two functions (`Scope::defer_free` and `Scope::defer_drop`).
+
+There might be very small differences in performance when using `defer_free` and `defer_drop` versus
```

Comment:
I think @schets may have an opinion on the performance.  cf: https://github.com/crossbeam-rs/crossbeam-epoch/pull/3#discussion_r129333625

Reply:
Good point.